R/model.R

```diff
@@ -914,6 +914,180 @@ sample_method <- function(data = NULL,
+#'     chain_ids = seq_len(chains),
+#'     iter_warmup = NULL,
+#'     iter_sampling = NULL,
+#'     save_warmup = FALSE,
+#'     thin = NULL,
+#'     max_treedepth = NULL,
+#'     adapt_engaged = TRUE,
+#'     adapt_delta = NULL,
+#'     step_size = NULL,
+#'     metric = NULL,
+#'     metric_file = NULL,
+#'     inv_metric = NULL,
+#'     init_buffer = NULL,
+#'     term_buffer = NULL,
+#'     window = NULL,
+#'     fixed_param = FALSE,
+#'     sig_figs = NULL,
+#'     validate_csv = TRUE,
+#'     show_messages = TRUE
+#'   )
+#'   ```
+#'
+#' @section Arguments:
+#'   * `mpi_cmd`: (character vector) The MPI launcher used for launching MPI processes.
+#'     The default launcher is `mpiexec`.
+#'   * `mpi_args`: (list) A list of arguments to use when launching MPI processes.
+#'     For example, mpi_args = list("n" = 4) launches the executable as
+#'     `mpiexec -n 4 model_executable`, followed by CmdStan arguments
+#'     for the model executable.
+#'   * `data`, `seed`, `refresh`, `init`, `save_latent_dynamics`, `output_dir`,
+#'     `chains`, `parallel_chains`, `chain_ids`, `iter_warmup`, `iter_sampling`,
+#'     `save_warmup`, `thin`, `max_treedepth`, `adapt_engaged`, `adapt_delta`,
+#'     `step_size`, `metric`, `metric_file`, `inv_metric`, `init_buffer`,
+#'     `term_buffer`, `window`, `fixed_param`, `sig_figs`, `validate_csv`,
+#'     `show_messages`:
+#'      Same as for the [`$sample()`][model-method-sample] method.
+#'
+#' @section Value: The `$mpi_sample()` method returns a [`CmdStanMCMC`] object.
+#'
+#' @template seealso-docs
+#' @inherit cmdstan_model examples
+#'
+NULL
+mpi_sample_method <- function(data = NULL,
+                          mpi_cmd = "mpiexec",
+                          mpi_args = NULL,
+                          seed = NULL,
+                          refresh = NULL,
+                          init = NULL,
+                          save_latent_dynamics = FALSE,
+                          output_dir = NULL,
+                          chains = 1,
+                          parallel_chains = getOption("mc.cores", 1),
+                          chain_ids = seq_len(chains),
+                          iter_warmup = NULL,
+                          iter_sampling = NULL,
+                          save_warmup = FALSE,
+                          thin = NULL,
+                          max_treedepth = NULL,
+                          adapt_engaged = TRUE,
+                          adapt_delta = NULL,
+                          step_size = NULL,
+                          metric = NULL,
+                          metric_file = NULL,
+                          inv_metric = NULL,
+                          init_buffer = NULL,
+                          term_buffer = NULL,
+                          window = NULL,
+                          fixed_param = FALSE,
+                          sig_figs = NULL,
+                          validate_csv = TRUE,
+                          show_messages = TRUE) {
```

Comment:
`threads_per_chain` is not an arg for `$mpi_sample()`, because if by any chance someone uses threading with MPIm `threads_per_chain` will always be 1, regardless of what is set.

Reply:
It's not unusual to use use threading in an MPI process, it's just we haven't done that for Stan.

Reply:
Agreed. If we enable it Stan, we can then add threading here also.